src/main/java/net/minestom/server/instance/block/predicate/DataComponentPredicate.java

```diff
@@ -0,0 +1,376 @@
+            EnchantmentList holderEnchants = holder.get(DataComponents.ENCHANTMENTS);
+            if (holderEnchants == null) {
+                return false;
+            }
+            if (enchantments == null) {
+                // If `enchantments` is not specified, the predicate returns true when any enchantment matches the specified `level`
+                return holderEnchants.enchantments().entrySet().stream().anyMatch(entry -> levels == null || levels.inRange(entry.getValue()));
+            }
+            for (DynamicRegistry.Key<net.minestom.server.item.enchant.Enchantment> e : enchantments) {
+                if (holderEnchants.has(e) && (levels == null || levels.inRange(holderEnchants.level(e)))) {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+    }
+
+    record StoredEnchantments(List<StoredEnchantment> children) implements DataComponentPredicate {
+        public static Codec<StoredEnchantments> CODEC = StoredEnchantment.CODEC.list().transform(StoredEnchantments::new, StoredEnchantments::children);
+
+        @Override
+        public boolean test(DataComponent.Holder holder) {
+            return children.stream().allMatch(enchantment -> enchantment.test(holder));
+        }
+    }
+
+    record StoredEnchantment(
+            @Nullable List<DynamicRegistry.Key<net.minestom.server.item.enchant.Enchantment>> enchantments,
+            Range.Int levels) {
+
+        public static Codec<StoredEnchantment> CODEC = StructCodec.struct(
+                "enchantments", net.minestom.server.item.enchant.Enchantment.CODEC.listOrSingle().optional(), StoredEnchantment::enchantments,
+                "levels", DataComponentPredicates.INT_RANGE_CODEC.optional(), StoredEnchantment::levels,
+                StoredEnchantment::new
+        );
+
+        public boolean test(DataComponent.Holder holder) {
+            EnchantmentList holderEnchants = holder.get(DataComponents.STORED_ENCHANTMENTS);
+            if (holderEnchants == null) {
+                return false;
+            }
+            if (enchantments == null) {
+                // If `enchantments` is not specified, the predicate returns true when any enchantment matches the specified `level`
+                return holderEnchants.enchantments().entrySet().stream().anyMatch(entry -> levels == null || levels.inRange(entry.getValue()));
+            }
+            for (DynamicRegistry.Key<net.minestom.server.item.enchant.Enchantment> e : enchantments) {
+                if (holderEnchants.has(e) && (levels == null || levels.inRange(holderEnchants.level(e)))) {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+    }
+
+    record Potions(List<PotionType> potionTypes) implements DataComponentPredicate {
+        public static Codec<Potions> CODEC = PotionType.CODEC.listOrSingle().transform(Potions::new, Potions::potionTypes);
```

Comment:
This doesn't handle the case of there being a tag of potion types. Essentially, per [the wiki](https://minecraft.wiki/w/Data_component_predicate#potion_contents) (and I did confirm otherwise that this is how it actually works), you need to be able to support a tag of something, e.g. for items, if the string `#minecraft:logs` is passed in it needs to refer to all logs. You also need to handle this in many other cases across the `DataComponent` predicates but I'm just mentioning this here so I don't have to spam it everywhere.

Reply:
It looks like this isn't possible for the `jukebox_playable` and `trim` predicates until the tag rework PR #2709 is merged, but I added tag support for `enchantments`, `stored_enchantments`, and `potion_contents`.